presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestJoinNodeFlattener.java

```diff
@@ -0,0 +1,367 @@
+        MultiJoinNode expected = new MultiJoinNodeBuilder()
+                .withSources(valuesA, valuesB, valuesC, valuesD, valuesE)
+                .withFilter(and(equals(a1, b1), equals(a1, c1), equals(d1, e1), equals(d2, e2), equals(b1, e1)))
+                .withOutputSymbols(a1, b1, c1, d1, d2, e1, e2)
+                .build();
+        assertEquals(toMultiJoinNode(joinNode, noLookup(), 5), expected);
+    }
+
+    @Test
+    public void testMoreThanJoinLimit()
+    {
+        PlanBuilder p = planBuilder();
+        Symbol a1 = p.symbol("A1");
+        Symbol b1 = p.symbol("B1");
+        Symbol c1 = p.symbol("C1");
+        Symbol d1 = p.symbol("D1");
+        Symbol d2 = p.symbol("D2");
+        Symbol e1 = p.symbol("E1");
+        Symbol e2 = p.symbol("E2");
+        ValuesNode valuesA = p.values(a1);
+        ValuesNode valuesB = p.values(b1);
+        ValuesNode valuesC = p.values(c1);
+        ValuesNode valuesD = p.values(d1, d2);
+        ValuesNode valuesE = p.values(e1, e2);
+        JoinNode join1 = p.join(
+                INNER,
+                valuesA,
+                valuesB,
+                ImmutableList.of(equiJoinClause(a1, b1)),
+                ImmutableList.of(a1, b1),
+                Optional.empty());
+        JoinNode join2 = p.join(
+                INNER,
+                valuesD,
+                valuesE,
+                ImmutableList.of(
+                        equiJoinClause(d1, e1),
+                        equiJoinClause(d2, e2)),
+                ImmutableList.of(
+                        d1,
+                        d2,
+                        e1,
+                        e2),
+                Optional.empty());
+        JoinNode joinNode = p.join(
+                INNER,
+                p.join(
+                        INNER,
+                        join1,
+                        valuesC,
+                        ImmutableList.of(equiJoinClause(a1, c1)),
+                        ImmutableList.of(a1, b1, c1),
+                        Optional.empty()),
+                join2,
+                ImmutableList.of(equiJoinClause(b1, e1)),
+                ImmutableList.of(
+                        a1,
+                        b1,
+                        c1,
+                        d1,
+                        d2,
+                        e1,
+                        e2),
+                Optional.empty());
+        MultiJoinNode expected = new MultiJoinNodeBuilder()
+                .withSources(join1, join2, valuesC)
+                .withFilter(and(equals(a1, c1), equals(b1, e1)))
+                .withOutputSymbols(a1, b1, c1, d1, d2, e1, e2)
+                .build();
+        assertEquals(toMultiJoinNode(joinNode, noLookup(), 2), expected);
+    }
+
+    private ComparisonExpression equals(Symbol left, Symbol right)
+    {
+        return new ComparisonExpression(EQUAL, left.toSymbolReference(), right.toSymbolReference());
+    }
+
+    private static final class MultiJoinNodeBuilder
```

Comment:
Move it to the `MultiJoinNode`. Call it `Builder`. It might be needed at some point in some other test. We don't want to have it implemented twice. Add a `MultiJoinNode.builder()` convenience method as we do for the other classes.